File: Algorithms/p206_Reverse_Linked_List_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head && head->next){
            ListNode* p = reverseList(head->next);
            head->next->next = head;
            head->next = 0;
            return p;
        }else{
            return head;
        }
    }
};

File: Algorithms/p206_Reverse_Linked_List_2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head and head.next:
            p = self.reverseList(head.next)
            head.next.next = head
            head.next = None
            return p
        else:
            return head
